### ISSUE TITLE
Shipping Address May Incorrectly Be Used in Cart

### ISSUE THREAD

**Issue by barryhughes** [woocommerce/woocommerce#29744]
Spun out of an observation made in [#28759](https://github.com/woocommerce/woocommerce/issues/28759#issuecomment-815295121):

- Suppose `Settings ‣ Shipping Options ‣ Shipping Destination` is set to `Default to customer billing address`
- And we have a logged-in customer who has updated their account with a saved billing address and a saved shipping address
- When they add a shippable item to the cart, the cart page will show their *shipping* address as the shipping address (contrary to the above setting, which should mean the billing address is used)
- On proceeding to the checkout the problem is corrected

**Prerequisites (mark completed items with an [x]):**
- [x] I have have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Expected behavior**

If `Settings ‣ Shipping Options ‣ Shipping Destination` is set to `Default to customer billing address` then the saved **billing address** (and not the saved shipping address) should be used as the shipping destination unless the customer manually decides to specify a different address.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Comment by juliaamosova** [comment]
Thank you for filing the issue @barryhughes! I assigned it a `high` priority since it affects the default functionality of WooCommerce that should be working flawlessly.

**Comment by barryhughes** [comment]
There is another facet to this. I was umming and ahhing about splitting it out into its own issue, but it _probably_ belongs here as part of this one:

1. Same setup as above.
2. Once on the cart page, actually change the shipping address and supply something completely new (distinct from either of the saved addresses). Lock it in by clicking _Update._
3. Now proceed to checkout:
    1. Though we just specified a separate shipping address, the `Ship to a different address?` toggle is **off** and unless we turn it on, the order will ship to the customer's billing address.
    2. If we do turn the toggle on, our new shipping address is surfaced: we can then complete any missing required fields (like the street address), place the order and it will be used.

@juliaamosova just flagging for you to review...are you happy if we keep these problems in this one issue (I'm thinking we'll need to address them together), or prefer they be separated out?

**Comment by vedanshujain** [comment]
@barryhughes I think this is intended, the setting `Shipping Destination` is used to only control the checkbox `Ship to a different address` in the checkout box. If a customer has explicitly set a shipping address then it should be used, instead of copying the billing address (although the button, ship to same address should be selected by default, which looks like it is).

Second issue seems something we can address, but I am inclined to reduce the priority here. WDYT?

**Comment by barryhughes** [comment]
@vedanshujain—you mean what I described as the primary issue is intended? Could be ... but if so I think we still have a poor UX in this regard. For instance, suppose I proceed to the cart. I might see something like this (note the shipping fees):

<img width="829" alt="cart" src="https://user-images.githubusercontent.com/3594411/121069083-eca6e000-c781-11eb-876f-eabea21f3d44.png">

Then if I do nothing else except proceed to the checkout, and I might see something like this (note the shipping fees have changed):

<img width="856" alt="checkout" src="https://user-images.githubusercontent.com/3594411/121069205-14964380-c782-11eb-915e-69b4aeefd3ec.png">

So we have an element of "whip-lash" for the customer—all they did was click on _Proceed to Checkout_ yet suddenly the shipping costs and order total have changed. In my example the shipping costs went down, but they could just as easily go up (and take the order total up with them).

It just doesn't feel smooth to me and perhaps it could even come across as untrustworthy: at least, I know that if I was buying from an online store I hadn't frequented before, and hadn't yet built up a sense of trust, and on proceeding to checkout suddenly and inexplicably (from my perspective as the customer) the fees suddenly jacked up, I would raise an eyebrow. 

tl;dr— I still think the primary issue is worth addressing. I'm less sure of the priority, probably it can be lowered.

**Comment by ralucaStan** [comment]
@nielslange could you test if this issue still applies? It wasn't triaged by someone in our team.

**Comment by nielslange** [comment]
> @nielslange could you test if this issue still applies? It wasn't triaged by someone in our team.

Sure. I'll test it now, @ralucaStan.

**Comment by nielslange** [comment]
@ralucaStan I just finished my investigation, and could reproduce the issue, mentioned initially. I've also tested this case against the Cart and Checkout blocks, as well as mixed scenarios. Below, you can find the various scenarios test cases. Because I could reproduce the issue, I moved it from `Needs triage` to `Backlog`.

## Test scenarios

<details>
<summary>Before each test</summary>
<br>

1. Go to `/wp-admin/profile.ph[p](https://store.test/wp-admin/profile.php)`.
2. Set the **customer billing address** to:

<table border="1">
  <tr>
    <th>Field</th>
    <th>Value</th>
  </tr>
  <tr>
    <td>Address line 1</td>
    <td>Stationsplein 23</td>
  </tr>
  <tr>
    <td>City</td>
    <td>Amsterdam</td>
  </tr>
  <tr>
    <td>Postcode / ZIP</td>
    <td>1010 AA</td>
  </tr>
  <tr>
    <td>Country / Region</td>
    <td>Netherlands</td>
  </tr>
  <tr>
    <td>Phone</td>
    <td>020-123456789</td>
  </tr>
</table>

3. Set the **customer shipping address** to:

<table border="1">
  <tr>
    <th>Field</th>
    <th>Value</th>
  </tr>
  <tr>
    <td>Address line 1</td>
    <td>Brandarisstraat 2</td>
  </tr>
  <tr>
    <td>City</td>
    <td>West-Terschelling</td>
  </tr>
  <tr>
    <td>Postcode / ZIP</td>
    <td>8881 AW</td>
  </tr>
  <tr>
    <td>Country / Region</td>
    <td>Netherlands</td>
  </tr>
  <tr>
    <td>Phone</td>
    <td>088-123456789</td>
  </tr>
</table>

4. Go to `wp-admin/admin.php?page=wc-settings&tab=shipping&section=options`.
5. As `Shipping destination` select the option `Default to customer billing address`
6. Go to the frontend and add a product to the cart.
<hr>
</details>

<details>
<summary>Test 1: Cart and Checkout blocks</summary>
<br>

1. Follow the steps as mentioned in `Before each test`. 
2. Go to the cart page with the Cart block.
3. 🔴 See that the shipping destination, in the sidebar, says `Shipping to 8881 AW, West-Terschelling, Netherlands` instead of `Shipping to 1010 AA, Amsterdam, Netherlands`.
4. Go to the checkout page with the Checkout block.
5. 🔴 See that the shipping destination, in the sidebar, says `Shipping to 8881 AW, West-Terschelling, Netherlands` instead of `Shipping to 1010 AA, Amsterdam, Netherlands`.
6. 🔴 See that the shipping destination, in the main column, says `Brandarisstraat 2, 8881 AW West-Terschelling, Netherlands, 088-123456789` instead of `Stationsplein 23, 1010 AA Amsterdam, Netherlands, 020-123456789`.
8. Go back to the cart page.
9. 🔴 See that the shipping destination, in the sidebar, still says `Shipping to 8881 AW, West-Terschelling, Netherlands`.
10. Go back to the checkout page.
11. Place the order.
12. 🔴 See that the shipping address is `Brandarisstraat 2, 8881 AW West-Terschelling, Netherlands, 088-123456789`, while the billing address is `Stationsplein 23, 1010 AA Amsterdam, Netherlands, 020-123456789`.
<hr>
</details>

<details>
<summary>Test 2: Cart and checkout shortcode</summary>
<br>

1. Follow the steps as mentioned in `Before each test`. 
2. Go to the cart page with the cart shortcode.
3. 🔴 See that the shipping destination, in the cart totals section, says `Shipping to **Brandarisstraat 2, 8881 AW West-Terschelling, Netherlands**.` instead of `Shipping to **Stationsplein 23, 1010 AA Amsterdam, Netherlands**.`
4. Go to the checkout page with the checkout shortcode.
5. 🟢 See that the checkbox for `Ship to a different address?` is unchecked, meaning the billing address is used as the shipping address.
6. Go back to the cart page.
7. 🟢 See that the shipping destination, in the cart totals section, says `Shipping to **Stationsplein 23, 1010 AA Amsterdam, Netherlands**.`
8. Go back to the checkout page.
9. 🟢 See that the checkbox for `Ship to a different address?` is still unchecked.
10. Place the order.
11. 🟠 See that both the shipping and the billing address are `Stationsplein 23, 1010 AA Amsterdam, Netherlands`. But note that only the billing address shows the phone number `020-123456789`.
13. Go to `/wp-admin/profile.ph[p](https://store.test/wp-admin/profile.php)`.
14. 🟠 See that the customer’s billing and shipping address are identical, but the phone number remained different.
<hr>
</details>

<details>
<summary>Test 3: Cart block and checkout shortcode</summary>
<br>

1. Follow the steps as mentioned in `Before each test`. 
2. Go to the cart page with the Cart block.
3. 🔴 See that the shipping destination, in the sidebar, says `Shipping to 8881 AW, West-Terschelling, Netherlands` instead of `Shipping to 1010 AA, Amsterdam, Netherlands`.
4. Go to the checkout page with the checkout shortcode.
5. 🟢 See that the checkbox for `Ship to a different address?` is unchecked, meaning the billing address is used as the shipping address.
6. Go back to the cart page.
7. 🟢 See that the shipping destination, in the cart totals section, says `Shipping to Stationsplein 23, 1010 AA Amsterdam, Netherlands.`
8. Go back to the checkout page.
9. 🟢 See that the checkbox for `Ship to a different address?` is still unchecked.
10. Place the order.
11. 🟠 See that both the shipping and the billing address are `Stationsplein 23, 1010 AA Amsterdam, Netherlands`. But note that only the billing address shows the phone number `020-123456789`.
13. Go to `/wp-admin/profile.ph[p](https://store.test/wp-admin/profile.php)`.
14. 🟠 See that customer’s billing and shipping address are identical, but the phone number remained different.
<hr>
</details>

<details>
<summary>Test 4: Cart shortcode and Checkout block</summary>
<br>

1. Follow the steps as mentioned in `Before each test`. 
2. Go to the cart page with the cart shortcode.
3. 🔴 See that the shipping destination, in the cart totals section, says `Shipping to **Brandarisstraat 2, 8881 AW West-Terschelling, Netherlands**.` instead of `Shipping to **Stationsplein 23, 1010 AA Amsterdam, Netherlands**.`
4. Go to the checkout page with the Checkout block.
5. 🔴 See that the shipping destination, in the sidebar, says `Shipping to 8881 AW, West-Terschelling, Netherlands` instead of `Shipping to 1010 AA, Amsterdam, Netherlands`.
6. 🔴 See that the shipping destination, in the main column, says `Brandarisstraat 2, 8881 AW West-Terschelling, Netherlands, 088-123456789` instead of `Stationsplein 23, 1010 AA Amsterdam, Netherlands, 020-123456789`.
7. Go back to the cart page.
8. 🔴 See that the shipping destination, in the cart totals section, still says `Shipping to **Brandarisstraat 2, 8881 AW West-Terschelling, Netherlands**.`.
9. Go back to the checkout page.
10.  🔴 See that the shipping destination, in the sidebar, still says `Shipping to 8881 AW, West-Terschelling, Netherlands`.
11. 🔴 See that the shipping destination, in the main column, still says `Brandarisstraat 2, 8881 AW West-Terschelling, Netherlands`.
12. Place the order.
13.  🔴 See that the shipping address is `Brandarisstraat 2, 8881 AW West-Terschelling, Netherlands, 088-123456789`, while the billing address is `Stationsplein 23, 1010 AA Amsterdam, Netherlands, 020-123456789`.
<hr>
</details>

## Expected results
- The billing address should be displayed as the shipping address when using the Cart block.
- The billing address should be displayed as the shipping address when using the Checkout block.
- The billing address should be displayed as the shipping address when using the cart shortcode.
- The billing address should be used as the shipping address when placing the order. (This fails for the tests 1 and 4.)
- The billing address should be displayed as the shipping address on the order confirmation page. (This fails for the tests 1 and 4.)

## Questions
- Is it expected that, when using the billing address as the shipping address, that the shipping address in `/wp-admin/profile.php` gets overwritten? If so, we should also overwrite the phone number.

## Suggestions
- Apart from addressing the identified issues, I suggest also adding e2e tests for early error detection.
